app/employee-app/backend/modules/database/db_functions.bal

```diff
@@ -0,0 +1,440 @@
+//
+// Copyright (c) 2005-2024, WSO2 LLC.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+// 
+import employee_app.types;
+
+import ballerina/log;
+import ballerina/sql;
+
+# Insert a new recruit.
+#
+# + recruit - Recruit data
+# + createdBy - Person who created the recruit
+# + return - Recruit id or error
+public isolated function addRecruit(AddRecruitPayload recruit, string createdBy) returns int|error {
+
+    sql:ExecutionResult|sql:Error executionResults = databaseClient->execute(addRecruitQuery(recruit, createdBy));
+    if executionResults is sql:Error {
+        string customError = string `An error occurred while adding the recruit!`;
+        log:printError(customError, executionResults);
+        return error(customError);
+    }
+    return executionResults.lastInsertId.ensureType();
+}
+
+# Retrieve career function data from HRIS.
+#
+# + filter - Criteria to filter the data  
+# + limit - Number of records to retrieve
+# + offset - Number of records to offset
+# + return - List of career functions
+public isolated function getCareerFunctions(CareerFunctionFilter filter, int 'limit, int offset)
+    returns CareerFunction[]|error {
+
+    CareerFunction[] careerFunctions = [];
+    stream<DBCareerFunction, sql:Error?> resultStream = databaseClient->query(getCareerFunctionsQuery(
+        filter = filter, 'limit = 'limit, offset = offset));
+    error? iterateError = from DBCareerFunction careerFunction in resultStream
+        do {
+            string resDesignations = careerFunction.designations ?: "[]";
+            Designation[]|error designations = resDesignations.fromJsonStringWithType();
```

Comment:
We can't do that in here as `careerFunction.designations` is an optional string(`string? designations;`). Since `fromJsonStringWithType()` needs a string to do the operation, set it like that(even the casting didn't work.

Reply:
What I meant was something like

```ballerina
string? resDesignations = careerFunction.designations;
Designation[]|error designations = 
    (resDesignations is ()) ? [] : resDesignations.fromJsonStringWithType();
```

Reply:
Yes, we can implement it that way. I was suggesting that without using a temporary variable, achieving that might be difficult. Thank you for the clarification!